### PR TITLE
Generalize Harmonic and FlatBotton restraints to be referenced to centroids

### LINKED ISSUE #736: Implement restraints referenced to center of geometries of atom selections
This is just a placeholder to remind us that we will need to implement variants of the `Harmonic` and `FlatBottom` restraints that use the centers of geometry of DSL-specified atom selections in order to carry out host-guest (and like T4 lysozyme) binding free energies for the paper.

We can use [`CustomCentroidBondForce`](http://docs.openmm.org/7.1.0/api-python/generated/simtk.openmm.openmm.CustomCentroidBondForce.html#simtk.openmm.openmm.CustomCentroidBondForce) for this purpose. For example, `Harmonic` would use something like this
```
# Harmonic restraint between two groups of atoms
num_groups = 2
force = openmm.CustomCentroidBondForce(num_groups, "(K/2)*distance(g1,g2)^2")
force.addGlobalParameter("K", K)
force.addGroup(particles1); # create group 0
force.addGroup(particles2); # create group 1
force.addBond([0,1], []); # add bond between groups 0 and 1
system.addForce(force)
```

### PR DESCRIPTION
**Pull request by andrrizzi**
Fix #736 . Should be ready for review if tests pass!

Instead of taking an integer to specify the receptor and ligand atoms, `RadiallySymmetricRestraint` now takes either a list of particles indices or a DSL expression + topology.

If the list of particles for receptor and ligand only has 1 index, `CustomBondForce` is used instead of `CustomCentroidBondForce`. This is because the latter supports only 64bit platforms. This way we can still run tests with a 32bit one. For example, on the Macbooks not equipped with an NVIDIA card, the OpenCL platform is 32bit.

### CODE REVIEW

File: Yank/restraints.py [hunk 1]
```diff
@@ -714,22 +787,24 @@ def _create_restraint_force(self, particle1, particle2):
...
-        force = openmm.CustomBondForce('lambda_restraints * ' + self._energy_function)
-        force.addGlobalParameter('lambda_restraints', 1.0)
+        if len(particles1) == 1 and len(particles2) == 1:
+            # CustomCentroidBondForce works only on 64bit platforms. When the
+            # restrained groups only have 1 particle, we can use the standard
+            # CustomBondForce so that we can support 32bit platforms too.
+            energy_function = self._energy_function.replace('distance(g1,g2)', 'r')
+            force = openmm.CustomBondForce('lambda_restraints * ' + energy_function)
+            force.addBond(particles1[0], particles2[0], parameter_values)
+        else:
```
Comment (Lnaden) [inline Yank/restraints.py]: Should we add an exception for 32-bit architecture, or at bare minimum a full documentation that you have to have 64bit if you want to use this feature for multiple atoms?

File: Yank/restraints.py [hunk 2]
```diff
@@ -714,22 +787,24 @@ def _create_restraint_force(self, particle1, particle2):
         parameter_names, parameter_values = zip(*self._bond_parameters.items())
 
         # Create bond force and lambda_restraints parameter to control it.
-        force = openmm.CustomBondForce('lambda_restraints * ' + self._energy_function)
-        force.addGlobalParameter('lambda_restraints', 1.0)
+        if len(particles1) == 1 and len(particles2) == 1:
+            # CustomCentroidBondForce works only on 64bit platforms. When the
```
Comment (Lnaden) [inline Yank/restraints.py]: Where in the docs does it say it only supports 64-bit platforms? And what is the exact error it throws if you use the wrong platform?

**Comment by Lnaden** [comment]
Just a couple comments, mostly just documentation changes requested. Can you also add in the auto-function linking to the docs once you merge the conflicting restraints file in? It should just be documentation clashes, not actual functionality ones.

File: Yank/restraints.py (hunk 2 above)
Comment (andrrizzi) [inline Yank/restraints.py]: I didn't find it in the docs, but it's in the OpenCL platform code ([here](https://github.com/pandegroup/openmm/blob/master/platforms/opencl/src/OpenCLKernels.cpp#L5054)).

File: Yank/restraints.py (hunk 1 above)
Comment (andrrizzi) [inline Yank/restraints.py]: OpenMM already raises a very clear exception if that's the case on `Context` creation.

**Comment by andrrizzi** [comment]
> add this support to the YAML syntax

I didn't touch the YAML syntax docs because it redirects to the API documentation to see the options taken by each type of restraint.

> Can you also add in the auto-function linking to the docs 

Will do!

**Comment by andrrizzi** [comment]
I'm also going to separate the `restrained_atoms` argument in the `Boresch` constructor into two `restrained_receptor_atoms` and `restrained_ligand_atoms` because I've just realized that it'll be problematic to specify SMIRKS regions in the future with a single parameter.

File: Yank/restraints.py (hunk 1 above)
Comment (andrrizzi) [inline Yank/restraints.py]: I'll add the 64bit thing to the docstrings!

**Comment by jchodera** [comment]
Thanks everyone --- this is great!

If we wanted to add a test to our validation suite, we could also simulate a host-guest system where we keep all lambdas at zero except for turning on the restraint lambdas over the course of the simulation. If we keep the box size fixed at the standard state volume, the standard state correction should exactly cancel the restraining free energy and we should get a zero free energy for that leg. (alternatively, if the box size is arbitrary, we would get the log of the ratio of box volumes)

**Comment by andrrizzi** [comment]
@Lnaden sorry, the last two commits need review again.

>  to add a test to our validation suite

Sounds good! I'll try to make that work when I'll have the host-guest systems ready.

File: Yank/restraints.py [hunk 3]
```diff
@@ -1476,7 +1516,9 @@ def determine_missing_parameters(self, thermodynamic_state, sampler_state, topog
                              'restraint parameters...'.format(attempt, MAX_ATTEMPTS))
 
                 # Randomly pick non-collinear atoms.
-                self.restrained_atoms = self._pick_restrained_atoms(sampler_state, topography)
+                restrained_atoms = self._pick_restrained_atoms(sampler_state, topography)
+                self.restrained_receptor_atoms = restrained_atoms[:3]
```
Comment (Lnaden) [inline Yank/restraints.py]: Do we want this to auto-repick both the receptor AND ligand atoms? What if the user specifies a series of receptor atoms they always want to use and dont care about the ligand atoms? This method will re-pick the receptor ones as well.

File: Yank/tests/test_restraints.py [hunk 4]
```diff
@@ -221,24 +221,40 @@ def test_partial_parametrization():
         ('Harmonic', dict(spring_constant=2.0*unit.kilojoule_per_mole/unit.nanometer**2,
                           restrained_receptor_atoms=[5])),
         ('FlatBottom', dict(well_radius=1.0*unit.angstrom, restrained_ligand_atoms=[130])),
-        ('Boresch', dict(restrained_atoms=[106, 123, 117, 130, 131, 136],
+        ('Boresch', dict(restrained_ligand_atoms=[130, 131, 136],
```
Comment (Lnaden) [inline Yank/tests/test_restraints.py]: Doesn't this test also need the "restrainted_receptor_atoms"? And if you are wanting to just randomly choose the receptor atoms as per how it is coded right now, it will also randomly rechoose the ligand atoms.

File: Yank/restraints.py (hunk 3 above)
Comment (andrrizzi) [inline Yank/restraints.py]: The `_pick_restrained_atoms` won't re-pick atoms that have been explicitly given, it basically just give return them as they are.

File: Yank/tests/test_restraints.py (hunk 4 above)
Comment (andrrizzi) [inline Yank/tests/test_restraints.py]: I did it on purpose to test that only receptor atoms are picked at random and that ligand atoms are left as they are.

File: Yank/restraints.py [hunk 5]
```diff
@@ -1564,46 +1737,73 @@ def _pick_restrained_atoms(self, sampler_state, topography):
...
-        heavy_ligand_atoms = set(topography.ligand_atoms) & heavy_atoms
-        heavy_receptor_atoms = set(topography.receptor_atoms) & heavy_atoms
         if len(heavy_receptor_atoms) < 3 or len(heavy_ligand_atoms) < 3:
             raise ValueError('There must be at least three heavy atoms in receptor_atoms '
                              '(# heavy {}) and ligand_atoms (# heavy {}).'.format(
                                      len(heavy_receptor_atoms), len(heavy_ligand_atoms)))
 
-        # Find valid pairs of ligand/receptor atoms within a cutoff
+        # If r3 or l1 atoms are given. We have to pick those.
+        if isinstance(heavy_receptor_atoms, list):
```
Comment (Lnaden) [inline Yank/restraints.py]: This will fail if they are a tuple or a set (which it will be if auto-chosen), which should be equally valid

File: Yank/restraints.py [hunk 6]
```diff
@@ -1564,46 +1737,73 @@ def _pick_restrained_atoms(self, sampler_state, topography):
...
+            return self.restrained_receptor_atoms + self.restrained_ligand_atoms
+
+        # If receptor and ligand atoms are explicitly provided, use those.
+        heavy_ligand_atoms = self.restrained_ligand_atoms
+        heavy_receptor_atoms = self.restrained_receptor_atoms
+
+        # Otherwise we restrain only heavy atoms.
+        heavy_atoms = set(topography.topology.select('not element H').tolist())
+        # Intersect heavy atoms with receptor/ligand atoms (s1&s2 is intersect).
+        if heavy_ligand_atoms is None:
```
Comment (Lnaden) [inline Yank/restraints.py]: Just noticed, if user chosen, hydrogens can get selected as the heavy atoms to use here, may be a problem.

File: Yank/restraints.py (hunk 5 above)
Comment (Lnaden) [inline Yank/restraints.py]: More exactly, this will always be False under many conditions unexpectedly, which may confuse our users

File: Yank/restraints.py (hunk 6 above)
Comment (andrrizzi) [inline Yank/restraints.py]: Are you suggesting to raise a warning? I wouldn't forbid the user to pick hydrogens if he wanted to (with mass repartitioning for example).

File: Yank/restraints.py (hunk 5 above)
Comment (andrrizzi) [inline Yank/restraints.py]: Shouldn't be a problem. `heavy_receptor_atoms` and `heavy_ligand_atoms` will be `list`s only if the user has specified them, otherwise they'll be `set`s.

File: Yank/restraints.py (hunk 6 above)
Comment (Lnaden) [inline Yank/restraints.py]: Hmm, good point, that is fine

File: Yank/restraints.py (hunk 5 above)
Comment (Lnaden) [inline Yank/restraints.py]: What is stopping the user from passing in a tuple or set of heavy atoms for either of those themselves?

Comment (andrrizzi) [inline Yank/restraints.py]: Ah, I see what you're saying now. The user can specify tuples, but the `restrained_receptor/ligand_atoms` property converts them to lists.

Comment (Lnaden) [inline Yank/restraints.py]: Ah, I see now. I've made a 3rd pass through and retract my concerns. lgtm

**Comment by andrrizzi** [comment]
Awesome, thanks!

**Comment by davidlmobley** [comment]
@elkhoury - looks like centroid restraints are in Yank now.

Thanks, all! @andrrizzi - would love sample inputs as soon as you have them (as I think Levi mentioned we're after seeing how to activate the different types of restraints in Yank in the context of a binding calculation, such as binding of a host to a guest -- flat bottom restraints, Boresch-style, and no restraints at all. Sample yaml files would be awesome).

We're going to be doing some benchmarking, initially on some specific host-guest system from the benchmark sets paper (probably overlapping with what you're looking at for the Yank paper) to examine the efficiency of different forms of restraints. Léa El Khoury, who I tagged above, is a new postdoc in the group just starting to figure out how to use Yank, so if you're able to share some examples as you set up these systems that will help her out. We're coordinating with Mehtap on this as well as the thought is that ultimately the work we're doing will connect up with her studies on multiple-site binding in HSA.

Thanks!

**Comment by jchodera** [comment]
> We're going to be doing some benchmarking, initially on some specific host-guest system from the benchmark sets paper (probably overlapping with what you're looking at for the Yank paper) to examine the efficiency of different forms of restraints

@andrrizzi is in the process of setting those exact systems up right now for the accuracy benchmark in the YANK paper.

**Comment by davidlmobley** [comment]
Yup. We'll use just one of them. I just wanted to point out the overlap since if he just lets us spectate on his setup then we'll be able to learn a lot. :)

**Comment by andrrizzi** [comment]
> so if you're able to share some examples as you set up these systems that will help her out.

Sure! I hope I'll be able to create examples in the next couple of days. I'll upload the files in a repo and ping you guys as soon as they are ready.
